crypto/libsodium-fork/src/libsodium/crypto_sign/ed25519/ref10/batch.c

```diff
@@ -0,0 +1,318 @@
+* implemention. we changed the algorithm according to https://eprint.iacr.org/2020/1244.pdf .
+* the batch size is between 3 and 64 sigantures per batch. 
+* When the batch fails the function falls back to check singature one at a time.
+* the function returns 0 on success and fills and array of "valid" ints. 
+* 1 - for signature i passed verification
+* 0 - for siganture i failed verification
+*/
+int crypto_sign_ed25519_open_batch(const unsigned char **m, unsigned long long *mlen, const unsigned char **pk, const unsigned char **RS, size_t num, int *valid)
+{
+    batch_heap batch;
+    ge25519_p3  p;
+    ge25519_p3  p_mul;
+    sc25519 *r_scalars;
+    size_t i, batchsize;
+    unsigned char hram[64];
+    unsigned char batchsum[32];
+    int ret = 0;
+
+    for (i = 0; i < num; i++)
+        valid[i] = 1;
+
+    while (num > 3) {
+        batchsize = (num > MAX_BATCH_SIZE) ? MAX_BATCH_SIZE : num;
+
+        /* valida the public key and siganture */
+        for (i=0; i < batchsize; i++){
+            if (validate_ed25519_pk_and_sig(RS[i],pk[i]) != 0)
+                goto fallback;
+        }
+
+        /* generate r (scalars[batchsize+1]..scalars[2*batchsize] */
+        ed25519_randombytes_unsafe (batch.r, batchsize * 16);
+        
+
+        r_scalars = &batch.scalars[batchsize + 1];
+        for (i = 0; i < batchsize; i++)
+            expand256_modm(r_scalars[i], batch.r[i], 16);
+
+        /* compute scalars[0] = ((r1s1 + r2s2 + ...)) */
+        for (i = 0; i < batchsize; i++) {
+            expand256_modm(batch.scalars[i], RS[i] + 32, 32);
+            mul256_modm(batch.scalars[i], batch.scalars[i], r_scalars[i]);
+        }
+        for (i = 1; i < batchsize; i++)
+            add256_modm(batch.scalars[0], batch.scalars[0], batch.scalars[i]);
+
+        /* compute scalars[1]..scalars[batchsize] as r[i]*H(R[i],A[i],m[i]) */
+        for (i = 0; i < batchsize; i++) {
+            crypto_hash_sha512_state hs;   
+            crypto_hash_sha512_init(&hs);         
+            crypto_hash_sha512_update(&hs, RS[i], 32);
+            crypto_hash_sha512_update(&hs, pk[i], 32);
+            crypto_hash_sha512_update(&hs, m[i], mlen[i]);
+            crypto_hash_sha512_final(&hs, hram);
+
+
+            expand256_modm(batch.scalars[i+1], hram, 64);
+            mul256_modm(batch.scalars[i+1], batch.scalars[i+1], r_scalars[i]);
+        }
+        /* compute points */
+        batch.points[0] = ge25519_basepoint;
+        for (i = 0; i < batchsize; i++)
+            if (ge25519_frombytes_negate_vartime(&batch.points[i+1], pk[i]) != 0)
+                goto fallback;
+        for (i = 0; i < batchsize; i++)
+            if (ge25519_frombytes_negate_vartime(&batch.points[batchsize+i+1], RS[i]) != 0)
+                goto fallback;
+
+        ge25519_multi_scalarmult_vartime(&p, &batch, (batchsize * 2) + 1);
+
+        ge25519_mul_by_cofactor(&p_mul, &p);
```

Comment:
In diffing against the libdonna implementation this is the only significant difference... how did it work before in libdonna? I thought you had to use the cofactor equation for batch verification.

Reply:
if I recall correctly, libdonna didn't use the cofactor version. that is one of the reasons why we need to change the implementation

Reply:
right OK! I now understand better that this was part of the Tor issue https://gitlab.torproject.org/tpo/core/tor/-/issues/40078